classes/core/PKPRouter.inc.php

```diff
@@ -220,16 +220,17 @@ function getRequestedContextPath($request, $requestedContextLevel = 1) {
 	 * A Generic call to a context defining object (e.g. a Press, a Conference, or a SchedConf)
 	 * @param $request PKPRequest the request to be routed
 	 * @param $requestedContextLevel int (optional) the desired context level
+	 * @param $forceReload bool (optional) Force context to be pulled from db
 	 * @return object
 	 */
-	function &getContext($request, $requestedContextLevel = 1) {
+	function &getContext($request, $requestedContextLevel = 1, $forceReload = false) {
```

Comment:
I don't see `$forceReload` used -- where is this needed?

Reply:
:man_shrugging: My first guess was that it was addressing a function signature PHP notice, but I can't remember and I also can't find any uses.

Reply:
Removed.

Reply:
Ah! I found it! https://github.com/NateWr/pkp-lib/blob/i3594_forms/classes/core/Dispatcher.inc.php#L131-L138

So, since the schema determines the properties of the context, we have to force reload the context after we have loaded generic plugins. That way a generic plugin can modify the schema, and these changes are reflected in the context that is attached to `$request` everywhere.

I don't _think_ that we can load plugins before loading the context, since we have to know which plugins should be enabled. But I'm open to alternative ideas if you don't like this approach.

Reply:
Agreed, we can't load the plugins before loading the context. But could the plugins, as they're loaded, alter the in-memory `Context` object?

Reply:
I think it would be possible but difficult, without much performance gain.

The schema defines the properties of an object, as well as how those properties should be validated and transformed going in/out of the database. A plugin may change the schema in order to add/edit/remove a property.

So it would be possible to inspect the schema before/after plugins are loaded, identify differences, pull any new or edited properties out of the database, and update the context with the new data while removing any properties that are no longer present in the schema.

This is pretty close to what it does now, except that the `Context` object is completely replaced, rather than just updating the `Context::$_data` property.

My main concern with trying to only update changed properties is that the schema definitions for objects and arrays can get complicated, with nested objects and arrays, and default content. (See [submissionChecklist](https://github.com/NateWr/pkp-lib/blob/i3594_forms/schemas/context.json#L588-L641).) PHP's object comparison may be robust enough to solve this, but I worry about running into missed changes that are hard to diagnose when they're this deep in the system.

If you think the performance gains are worth it, I can work on special handling for refreshing the `Context` and `Site` objects, since I think these are the only two which would be loaded before plugins.

Reply:
No, I don't think it's worth the effort -- let's continue as you've got it.